framework/src/main/java/org/apache/felix/framework/util/ThreadGate.java

```diff
@@ -72,14 +74,14 @@ public synchronized void setMessage(Object msg)
     **/
     public synchronized boolean await(long timeout) throws InterruptedException
     {
-        long start = System.currentTimeMillis();
+        long start = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
```

Comment:
Could convert the timeout input to nanoseconds instead of converting system.nanoTime to millis possibly multiple times.